packages/bsky/src/db/migrations/20231003T202833377Z-create-moderation-subject-status.ts

```diff
@@ -0,0 +1,116 @@
+import { Kysely } from 'kysely'
+
+export async function up(db: Kysely<unknown>): Promise<void> {
+  await db.schema
+    .createTable('moderation_event')
+    .addColumn('id', 'serial', (col) => col.primaryKey())
+    .addColumn('action', 'varchar', (col) => col.notNull())
+    .addColumn('subjectType', 'varchar', (col) => col.notNull())
+    .addColumn('subjectDid', 'varchar', (col) => col.notNull())
+    .addColumn('subjectUri', 'varchar')
+    .addColumn('subjectCid', 'varchar')
+    .addColumn('comment', 'text')
+    .addColumn('meta', 'jsonb')
+    .addColumn('createdAt', 'varchar', (col) => col.notNull())
+    .addColumn('createdBy', 'varchar', (col) => col.notNull())
+    .addColumn('reversedAt', 'varchar')
+    .addColumn('reversedBy', 'varchar')
+    .addColumn('durationInHours', 'integer')
+    .addColumn('expiresAt', 'varchar')
+    .addColumn('reversedReason', 'text')
+    .addColumn('createLabelVals', 'varchar')
+    .addColumn('negateLabelVals', 'varchar')
+    .addColumn('legacyRefId', 'integer')
+    .execute()
+  await db.schema
+    .createTable('moderation_subject_status')
+    .addColumn('id', 'serial', (col) => col.primaryKey())
+
+    // Identifiers
+    .addColumn('did', 'varchar', (col) => col.notNull())
+    // Default to '' so that we can apply unique constraints on did and recordPath columns
+    .addColumn('recordPath', 'varchar', (col) => col.notNull().defaultTo(''))
+    .addColumn('blobCids', 'jsonb')
+    .addColumn('recordCid', 'varchar')
+
+    // human review team state
+    .addColumn('reviewState', 'varchar', (col) => col.notNull())
+    .addColumn('comment', 'varchar')
+    .addColumn('muteUntil', 'varchar')
+    .addColumn('lastReviewedAt', 'varchar')
+    .addColumn('lastReviewedBy', 'varchar')
+
+    // report state
+    .addColumn('lastReportedAt', 'varchar')
+
+    // visibility/intervention state
+    .addColumn('takendown', 'boolean', (col) => col.defaultTo(false).notNull())
+    .addColumn('suspendUntil', 'varchar')
+
+    // timestamps
+    .addColumn('createdAt', 'varchar', (col) => col.notNull())
+    .addColumn('updatedAt', 'varchar', (col) => col.notNull())
+    .addUniqueConstraint('moderation_status_unique_idx', ['did', 'recordPath'])
+    .execute()
+
+  // Move foreign keys from moderation_action to moderation_event
+  await db.schema
+    .alterTable('record')
+    .dropConstraint('record_takedown_id_fkey')
+    .execute()
+  await db.schema
+    .alterTable('actor')
+    .dropConstraint('actor_takedown_id_fkey')
+    .execute()
+  await db.schema
+    .alterTable('actor')
+    .addForeignKeyConstraint(
+      'actor_takedown_id_fkey',
+      ['takedownId'],
+      'moderation_event',
+      ['id'],
+    )
+    .execute()
+  await db.schema
+    .alterTable('record')
+    .addForeignKeyConstraint(
+      'record_takedown_id_fkey',
+      ['takedownId'],
+      'moderation_event',
+      ['id'],
+    )
+    .execute()
```

Comment:
I didn't see anything in the migration that would adjust these columns on `record` and `actor` while the `takedownId` will change, so this foreign key will become invalid.

One idea I have is to preserve the action->event ids, but not bother maintaining the report ids.  The process would be something like:
1. Insert all actions into the event table explicitly setting the `id` columns to match.
2. Update the next autoincrement event `id` value to be higher than that last id (maybe plus some gap) using `setval()`.
3. Insert reports into the event table using the autoincrementing id.

Reply:
Problem with re-sequencing report ids is that the application logic expects the ids to be sequential and pagination depends on that too so if we were to keep action ids matching their original ids but not reports, when looking at these data, it will always look like some actions were taken before reports came in since the sorting uses ids.

I am wondering if we even need this reference to events? isn't the subject_status table maintaining the real state anyways? if we really do need the reference, how about we update the takedownId in these tables to match the newly created event ids during migration?

Reply:
> Problem with re-sequencing report ids is that the application logic expects the ids to be sequential 

Ohh, I didn't notice that!  Where does that happen?

> I am wondering if we even need this reference to events? isn't the subject_status table maintaining the real state anyways?

It's a good point, `subject_status` is a new table to maintain this state and the `takedownId` column wasn't designed with it in mind.  However, I don't think that we can afford to join over to `subject_status` everywhere that we need to check for takedowns, so I expect that some state will remain on the actor/record rows to keep these checks very cheap.  If that is the case I think it's nearly all upside to maintain not just a boolean but a full id, which gives us better debuggability and helps rather than hurts with data integrity issues that can arise from having duplicated state.

> if we really do need the reference, how about we update the takedownId in these tables to match the newly created event ids during migration?

That works!  I think it would be ideal if the ids didn't change, but if they must change then we can update the `takedownId` column.  Part of the reason it would be nice not to touch that column is that it's quite sensitive, so we should just take some care to ensure that there's not a way to unset it (e.g. if we can't find a matching id for a given subject we should rollback rather than null it out).  On the pds we moved `takedownId` to a new column named `takedownRef`, and we could totally do the same thing here if it's easier to update the ids in a new column rather than in-place.


Overall I'd be interested to see where the id ordering matters in the application.  If it's not difficult to loosen that constraint it might be a nice outcome.  I am sure that we can handle the pagination issue by paginating on `(createdAt, id)` as we do many places.  If this ends-up not being very easy to do then we should look at updating the `takedownId` state.